File: BeaconActive-ObjC/BeaconActive-ObjC/SENLocationManager.m
//
//  SENLocationManager.m
//  BeaconActive-ObjC
//
//  Created by skyming on 15/4/10.
//  Copyright (c) 2015年 Sensoro. All rights reserved.
//

#import "SENLocationManager.h"
#import <UIKit/UIKit.h>

@interface SENLocationManager ()<CLLocationManagerDelegate>

@end

@implementation SENLocationManager

+ (instancetype)sharedInstance
{
    static dispatch_once_t once;
    static id sharedInstance;
    dispatch_once(&once, ^{
        sharedInstance = [[self alloc] init];
    });
    return sharedInstance;
}

- (instancetype)init{
    
    if(self = [super init])
    {
        _locationManager = [[CLLocationManager alloc]init];

        if ([[[UIDevice currentDevice] systemVersion] floatValue] >= 8.0)
        {
            [_locationManager requestAlwaysAuthorization];
        }
        _locationManager.delegate = self;
        _monitorRegion = [[CLBeaconRegion alloc]initWithProximityUUID:[[NSUUID alloc] initWithUUIDString:@"46D06053-9FAD-483B-B704-E576735CE1A3"] identifier:@"SensoroBeaconActiveTest"];
                          
    }
    return self;
}


- (void)startMonitor:(BOOL)relaunch{
    
    if (!relaunch) {
        [_locationManager startMonitoringForRegion:_monitorRegion];
        NSLog(@"Start monitor region!");
    }else{
        NSLog(@"During the relauch app, don't restart monitor region!");

    }
    _started = true;

}

- (void)stopMonitor{
    [_locationManager stopMonitoringForRegion:_monitorRegion];
    _started = false;
    NSLog(@"Stop monitor region!");
}

- (void)sendNotification:(NSString *)notify{
    UILocalNotification *notification = [[UILocalNotification alloc]init];
    notification.alertBody = notify;
    notification.soundName = UILocalNotificationDefaultSoundName;
    
    [[UIApplication sharedApplication] scheduleLocalNotification:notification];
}

- (void)locationManager:(CLLocationManager *)manager didDetermineState:(CLRegionState)state forRegion:(CLRegion *)region{
    
    NSDate *now = [NSDate date];
    NSDateFormatter *formatter = [[NSDateFormatter alloc]init];
    formatter.dateFormat = @"YYYY/MM/dd HH:mm:ss";
    
    switch (state) {
        case CLRegionStateInside:
            {
                NSString *alertbody =[NSString stringWithFormat:@"Enter region at %@",[formatter stringFromDate:now]];
                [self sendNotification:alertbody];
                NSLog(@"%@",alertbody);
            }
            break;
            
        case CLRegionStateOutside:
            {
                NSString *alertbody =[NSString stringWithFormat:@"Exit region at %@",[formatter stringFromDate:now]];
                [self sendNotification:alertbody];
                NSLog(@"%@",alertbody);
            }
        break;
    
        case CLRegionStateUnknown:
            NSLog(@"This region state was unknown!");
            break;
    }
}

- (void)locationManager:(CLLocationManager *)manager didFailWithError:(NSError *)error{
    NSLog(@"did Fail With Error %@", error.localizedDescription);
}

@end

File: BeaconActive-Swift/BeaconActive-Swift/SENLocationManager.swift
//
//  SENLocationManager.swift
//  BeaconActive-Swift
//
//  Created by David Yang on 15/4/3.
//  Copyright (c) 2015年 Sensoro. All rights reserved.
//

import UIKit
import CoreLocation

class SENLocationManager: NSObject, CLLocationManagerDelegate {

    static let sharedInstance = SENLocationManager();

    let locationManager = CLLocationManager();
    var started = false;
    var monitorRegion : CLBeaconRegion!;

    fileprivate override init(){
        
        super.init();
        if UIDevice.current.systemVersion.compare("8.0", options: .numeric, range: nil, locale: nil) != .orderedAscending {
            locationManager.requestAlwaysAuthorization();
        }
        locationManager.delegate = self;
        
        monitorRegion = CLBeaconRegion(proximityUUID: UUID(uuidString: "46D06053-9FAD-483B-B704-E576735CE1A3")!,
            identifier: "SensoroBeaconActiveTest");
        
//        monitorRegion.notifyOnEntry = true;
//        monitorRegion.notifyOnExit = true;
    }
    
    func startMonitor( relaunch : Bool ){
        if relaunch == false {
            locationManager.startMonitoring(for: monitorRegion);
            locationManager.startRangingBeacons(in: monitorRegion);
            NSLog("Start monitor region!");
        }else{
//            locationManager.startRangingBeacons(in: monitorRegion);
            NSLog("During the relauch app, don't restart monitor region!");
        }
        started = true;
    }
    
    func stopMonitor(){
        locationManager.stopMonitoring(for: monitorRegion);
        started = false;
        NSLog("Stop monitor region!");
    }
    
    func sendNotification(notify : String){
        let notification = UILocalNotification()
        notification.alertBody = notify
        
//        notification.applicationIconBadgeNumber = UIApplication.sharedApplication().applicationIconBadgeNumber + 1
        notification.soundName = UILocalNotificationDefaultSoundName
        UIApplication.shared.scheduleLocalNotification(notification)
    }
    
    //MARK: CLLocationManagerDelegate
    func locationManager(_ manager: CLLocationManager, didDetermineState state: CLRegionState, for region: CLRegion) {
        
        switch state {
        case .inside:
            let now = Date();
            let formatter = DateFormatter();
            formatter.dateFormat = "YYYY/MM/dd HH:mm:ss";
            sendNotification(notify: "Enter region at \(formatter.string(from: now))");
            NSLog("Enter region at \(formatter.string(from: now))");
        case .outside:
            let now = Date();
            let formatter = DateFormatter();
            formatter.dateFormat = "YYYY/MM/dd HH:mm:ss";
            sendNotification(notify: "Exit  region at \(formatter.string(from: now))");
            NSLog("Exit  region at \(formatter.string(from: now))");
        case .unknown:
            NSLog("This region state was unknown!");
        }
        
    }
    
    func locationManager(_ manager: CLLocationManager, didRangeBeacons beacons: [CLBeacon], in region: CLBeaconRegion) {
        print(#function);
        for beacon in beacons{
            print("beacon: \(beacon.description)")
        }
    }
    
    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        NSLog("did Fail With Error \(error.localizedDescription)");
    }
}
